File: scripts/blockstudio/bake-textured-mesh.mjs
#!/usr/bin/env node

/**
 * Node wrapper: bake textures onto a base mesh GLB via Blender.
 *
 * Invokes scripts/blockstudio/bake-textured-mesh.py in background Blender.
 * Pulls the job spec from a JSON file written ahead of time by the caller.
 *
 * Usage:
 *   node scripts/blockstudio/bake-textured-mesh.mjs --job /path/to/job.json
 *
 * The job JSON shape is defined in bake-textured-mesh.py.
 */

import fs from "node:fs";
import path from "node:path";
import { execFileSync } from "node:child_process";
import { fileURLToPath } from "node:url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const REPO_ROOT = path.resolve(__dirname, "..", "..");
const BLENDER = process.env.BLENDER_BIN || "/opt/homebrew/bin/blender";
const DRIVER = path.join(REPO_ROOT, "scripts/blockstudio/bake-textured-mesh.py");

function fatal(msg) {
  process.stderr.write(`[bake-textured-mesh] ${msg}\n`);
  process.exit(1);
}

/** Programmatic entry point — used by the api-proxy middleware. */
export function bakeTexturedMesh(jobPath) {
  if (!fs.existsSync(jobPath)) fatal(`job not found: ${jobPath}`);
  if (!fs.existsSync(BLENDER)) fatal(`Blender not found at ${BLENDER} (set BLENDER_BIN)`);

  execFileSync(
    BLENDER,
    [
      "--background",
      "--python", DRIVER,
      "--",
      "--job", jobPath,
    ],
    { stdio: "inherit" }
  );
}

function main() {
  const args = process.argv.slice(2);
  const jobIdx = args.indexOf("--job");
  if (jobIdx === -1 || !args[jobIdx + 1]) fatal("usage: bake-textured-mesh.mjs --job <path>");
  bakeTexturedMesh(args[jobIdx + 1]);
}

if (import.meta.url === `file://${process.argv[1]}`) {
  main();
}

File: scripts/blockstudio/bake-textured-mesh.py
"""
Standalone Blender driver: bake textures onto a base mesh GLB.

Takes a base mesh GLB (from assets/meshes/) plus a per-role texture set and
produces a new GLB with those textures applied, keyed by each mesh object's
`textureRole` custom property.

This is lighter weight than scripts/blockstudio/blender-rebuild-kit.py —
there's no planner, no kit spec, no procedural geometry. The base GLB is
authored upstream; we just swap materials.

Job JSON shape:
    {
      "baseMeshPath": "/abs/path/to/assets/meshes/wall.glb",
      "materials": {
        "wall":   { "baseColor": "/abs/path/...", "normal": "...", "arm": "...",
                    "roughnessFactor": 0.85, "metallicFactor": 0.0 },
        "trim":   { ... },
        "accent": { "synthetic": "glass" }
      },
      "outputPath": "/abs/path/to/assets/textured-meshes/<name>/artifact.glb"
    }

Usage:
    blender --background --python scripts/blockstudio/bake-textured-mesh.py -- \
        --job /path/to/job.json
"""

import sys
import os
import json
from argparse import ArgumentParser


def main():
    if "--" in sys.argv:
        argv = sys.argv[sys.argv.index("--") + 1:]
    else:
        argv = []

    parser = ArgumentParser(description="Bake textures onto a base mesh GLB.")
    parser.add_argument("--job", required=True, help="Path to job JSON.")
    args = parser.parse_args(argv)

    # Make the blender/ package importable (same trick as blender-rebuild-kit.py)
    repo_root = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
    blender_dir = os.path.join(repo_root, "blender")
    if blender_dir not in sys.path:
        sys.path.insert(0, blender_dir)

    import bpy
    from materials import apply_textures

    with open(args.job, "r", encoding="utf-8") as fh:
        job = json.load(fh)

    base_mesh_path = job["baseMeshPath"]
    materials = job.get("materials") or {}
    output_path = job["outputPath"]

    if not os.path.isfile(base_mesh_path):
        raise SystemExit(f"Base mesh not found: {base_mesh_path}")

    # Start from an empty scene
    bpy.ops.wm.read_factory_settings(use_empty=True)

    print(f"[bake-textured-mesh] Importing {base_mesh_path}")
    bpy.ops.import_scene.gltf(filepath=base_mesh_path)

    mesh_roles = [
        obj.get("textureRole")
        for obj in bpy.context.scene.objects
        if obj.type == "MESH" and obj.get("textureRole")
    ]
    print(f"[bake-textured-mesh] Imported meshes with roles: {mesh_roles}")

    if materials:
        roles_in_job = list(materials.keys())
        print(f"[bake-textured-mesh] Applying {len(materials)} material(s): {roles_in_job}")
        tex_result = apply_textures({"materials": materials})
        print(f"[bake-textured-mesh] Assigned material to {tex_result.get('assignedCount', 0)} meshes")
    else:
        print("[bake-textured-mesh] No materials provided — exporting bare")

    out_dir = os.path.dirname(output_path)
    os.makedirs(out_dir, exist_ok=True)

    print(f"[bake-textured-mesh] Exporting to {output_path}")
    bpy.ops.export_scene.gltf(
        filepath=output_path,
        export_format="GLB",
        export_extras=True,
        export_yup=True,
        export_apply=True,
    )
    print(f"[bake-textured-mesh] Done.")


if __name__ == "__main__":
    main()
